File: services/recommendation_service.py
"""
Recommendation Service
Generates trending songs and simple recommendations
"""
import time
from collections import Counter

class RecommendationService:
    @staticmethod
    def recommend(play_counts, user_stats):
        start = time.time()
        # Get top 5 songs by total plays
        trending = [k for k, _ in Counter(play_counts).most_common(5)]

        # Recommend trending songs to users who don't have them as favorites
        recommendations = {}
        for user in user_stats:
            fav = user["top_artist"]
            recommendations[user["user_id"]] = [
                s for s in trending if fav not in s
            ]

        processing_time = time.time() - start
        return {
            "trending_songs": trending,
            "recommendations": recommendations,
            "processing_time": processing_time
        }

File: grpc_mixed/go_service/recommendation_service.go
package services

import (
	"sort"
	"time"
)

type UserStats struct {
	UserID    string
	TopArtist string
}

type RecommendationResult struct {
	TrendingSongs   []string
	Recommendations map[string][]string
	ProcessingTime  float64
}

type RecommendationService struct{}

func (rs *RecommendationService) Recommend(playCounts map[string]int32, userStats []UserStats) RecommendationResult {
	start := time.Now()

	type kv struct {
		Key   string
		Value int32
	}
	var countSlice []kv
	for k, v := range playCounts {
		countSlice = append(countSlice, kv{k, v})
	}

	sort.Slice(countSlice, func(i, j int) bool {
		return countSlice[i].Value > countSlice[j].Value
	})

	trending := []string{}
	for i := 0; i < len(countSlice) && i < 5; i++ {
		trending = append(trending, countSlice[i].Key)
	}

	recommendations := make(map[string][]string)
	for _, u := range userStats {
		list := []string{}
		for _, song := range trending {
			if u.TopArtist != "" && !containsSubstring(song, u.TopArtist) {
				list = append(list, song)
			}
		}
		recommendations[u.UserID] = list
	}

	return RecommendationResult{
		TrendingSongs:   trending,
		Recommendations: recommendations,
		ProcessingTime:  time.Since(start).Seconds(),
	}
}

func containsSubstring(s, sub string) bool {
	return len(sub) > 0 && len(s) > 0 && Index(s, sub) != -1
}

func Index(str, substr string) int {
outer:
	for i := 0; i+len(substr) <= len(str); i++ {
		for j := range substr {
			if str[i+j] != substr[j] {
				continue outer
			}
		}
		return i
	}
	return -1
}
